### ISSUE TITLE
SRI Template Provider using ZMQ and RPC

### ISSUE THREAD

**Issue by Sjors** [stratum-mining/stratum#1032]
An [alternative approach](https://github.com/bitcoin/bitcoin/pull/29432#pullrequestreview-2156406658) for implementing the Template Provider was suggested during review.   

Rather than having Bitcoin Core accept incoming network connections, deal with Noise encryption and stratum messages, an external tool would do this. This tool would get its information from the node in two ways:

1. a new zmq publisher, e.g. -zmqpubtemplate, which publishes block templates as soon as they become available.
2. a new RPC `updatetemplatepub` to configure the template publisher, e.g. for setting the coinbase output data size

The external tool should probably be part of SRI, e.g. by adding `roles/template_provider`.

However, in this scenario it may not make sense to have that role in the first place. It just needs to be code shared by the The Job Declarator Client and the Pool roles.

It has similar implications as the idea of making bitcoind a Job Declarator Client, discussed in https://github.com/stratum-mining/sv2-spec/discussions/85. Except in that scenario we had to keep the Template Provider role around because the pool needs it, whereas here that role could disappear entirely. 

I could look into adding those two things in a new Bitcoin Core branch. But someone else would have to take on the SRI side.

Not sure if we actually want to go this route, but we can the discuss the pros and cons here.

**Comment by Fi3** [comment]
Having the additional role make sense also in this scenario, for example to allow public facing TPs.

**Comment by Fi3** [comment]
Going this route was my original plan and then we ruled it out (I'm trying hard to remember why but  I don't). Honestly to me seems that the advantages exceed the disadvantages. We don't even need to refactor the spec, and we will be much more flexible for example adding an extension that now need to be implemented in core and most likley be done in the TP role much faster

**Comment by Fi3** [comment]
@thebluematt what do you think?

**Comment by Fi3** [comment]
Another advantage of separeted TP is that the JDS don't need anymore to talk bitcoind rpc

**Comment by Sjors** [comment]
> for example to allow public facing TPs

Having a public facing TP won't work in this design, or at least you'd have to pick one `additional_coinbase_output_data_size` number (like 4000) for all clients (probably not too bad).

> I'm trying hard to remember why but I don't

Hopefully someone remembers...

**Comment by Sjors** [comment]
> JDS don't need anymore to talk bitcoind rpc

I think it does. The protocol between the JDC and JDS wouldn't change, the latter still needs to check against its mempool.

JDC, JDS and the pool (unless we keep the TP role) will need RPC access.

**Comment by Sjors** [comment]
Is there (good) Rust support for ZMQ?

**Comment by Fi3** [comment]
> > JDS don't need anymore to talk bitcoind rpc
> 
> I think it does. The protocol between the JDC and JDS wouldn't change, the latter still needs to check against its mempool.
> 
> JDC, JDS and the pool (unless we keep the TP role) will need RPC access.

yep I mean with TP role. It will be easier to implement a JDS (I expect pools to implement custom JDS not using the SRI one). We can just add an extension that do what now we are doing with RPC in the JDS.
